feat: rewards overhaul (#1876)
vegaprotocol/specs: protocol/0056-REWA-rewards_overview.md

- $\Delta{p}$ be the parties change in pnl in the ith epoch
- $N$ be the network parameter `rewards.metrics.relativeReturnsWindow`

$$m_{rr} = \frac{\sum_{i}^{n}{$\Delta{p}$\cdot\log(1 + \bar{OI_{i}})}}{N}$$

### Returns volatility metric

The return volatility metric, $m_{rv}$, measures the volatility of a parties returns across a number of epochs.

At the end of an epoch, if a party has had net returns less than or equal to `0` over the last $N$ epochs (where $N$ is the network parameter `rewards.metrics.returnsVolatilityWindow`), their reward metric $m_{rv}$ is set to `0`. Otherwise, the network should calculate the standard deviation of the set of each parties returns weighted by the log of their [time weighted average open interest](#open-interest-metric) over the last $N$ epochs.

Given the set:

$$R = \{r_i \cdot \log(1 +\bar{OI_{i}}) \mid i = 1, 2, \ldots, N\}$$

The reward metric $m_{rv}$ is the standard deviation of the set $R$.

### Validator power metric

The validator power metric, $m_{vp}$, measures the voting power of validators.

At the end of each epoch, for each party who is a validator set their reward metric as follows.

> $$m_{vp} = validatorScore$$

Reviewer: what validator score? why not just the voting power directly?

Reviewer: I generally don't understand how this is supposed to work - if I specify this metrics with no parties in scope, does anyone get it? the way it's written above, does it go only to consensus validators and not ersatz validators? pending validators? those would have non negative score (by some definition of score) but would have no voting power so not sure what the intention is. If I don't qualify parties, who gets the reward? no one? all validators ranked by their score?

If the reward account balance is 0 at the end of the epoch for a given market, any parties with non-zero metrics will not be rewarded for that epoch and their metric scores do not roll over (they are still zeroed).

Fee-based reward metrics (the total fees paid/received by each party as defined above) are stored in [LNL checkpoints](./0073-LIMN-limited_network_life.md) and are restored after a checkpoint restart to ensure rewards are not lost.

### Average position metric

The average position metric, $m_{ap}$, measures each parties time-weighted average position over a number of epochs.

At the start of each epoch, the network must reset each parties time weighted average position for the epoch ($\bar{P}$) to `0`. Whenever a parties position changes during an epoch, **and** at the end of the epoch, this value should be updated as follows.

Let:

- $\bar{P}$ be the parties time weighted average position in the epoch so far
- $P_{n}$ be the parties position before their position changed
- $t_{n}$ be the time the party held the previous position in seconds
- $t$ be the amount of time elapsed in the current epoch so far

> $$\bar{P} = \bar{P} \cdot \left(1 - \frac{t_{n}}{t}\right) + \frac{|P_{n}| \cdot t_{n}}{t}$$

Reviewer 2: ```suggestion
$$\bar{P}  \leftarrow  \bar{P} \cdot \left(1 - \frac{t_{n}}{t}\right) + \frac{|P_{n}| \cdot t_{n}}{t}$$
```

Reviewer 2: I suggest using $\leftarrow$ to indicate assignment - since \bar P is on both sides...

At the end of the epoch, the network must store the parties time weighted average position and then calculate their average position reward metric as follows.

Let:

- $m_{ap}$ be the parties average position reward metric

> - $\bar{P_{i}}$ be the parties time weighted average position in the i th epoch

Reviewer 2: ```suggestion
- $\bar{P_{i}}$ be the parties time weighted average position in the $i$-th epoch
```

- $N$ be the window length specified in the recurring transfer.

$$m_{ap} = \frac{\sum_{i}^{n}\bar{P_{i}}}{N}$$

### Relative return metric

The relative return metric, $m_{rr}$, measures each parties average relative return, weighted by their [time-weighted average position](#average-position-metric), over a number of epochs.

At the end of each epoch, the network must calculate and store the parties relative returns as follows.

Let:

- $r_i$ be the parties relative returns in the epoch
- $m2m_{wins}$ be the sum of all mark-to-market win transfers in the epoch
- $m2m_{losses}$ be the sum of all mark-to-market loss transfers in the epoch

> - $TWAP$ be the parties time-weighted average position in the epoch.

Reviewer 2: Any reason not to use $\bar P$ introduced above (instead of new notation)?

Author: ```suggestion
- $\bar{P}$ be the parties time-weighted average position in the epoch.
```

> $$r = \frac{m2m_{wins} + m2m_{losses}}{TWAP}$$

Reviewer 2: ```suggestion
$$r = \frac{m2m_{wins} - m2m_{losses}}{TWAP}$$
```

Reviewer 2: In fact we should probable clarify whether we expect $\text{m2m}_{\text{losses}}$ to come with a negative sign above...

Author: Good point, I think since it will be the sum of the "loss" transfers it should be a positive value. For extra clarity, though I will add abs signs to both.
```suggestion
$$r = \frac{|m2m_{wins}| - |m2m_{losses}|}{\bar{P}}$$
```

- $s_{i}$ be the share of the rewards for entity $i$

$$d_{i}=r_{i} M_{i}$$

Note if the entity is a team, $M_{i}$ is set to `1` as reward payout multipliers are considered later when distributing rewards [amongst the team members](#distributing-rewards-amongst-team-members).

Calculate each entities share of the rewards, $s_{i}$ pro-rata based on $d_{i}$, i.e.

$$s_{i} = \frac{d_{i}}{\sum_{i=1}^{n}d_{i}}$$

#### Distributing based on rank

Rewards funded using the rank-distribution strategy should be distributed as follows.

1. Calculate each entity's reward metric.
2. Arrange all entities in a list in descending order based on their reward metric values and determine their rank. If multiple entities share the same reward metric value, they should be assigned the same rank. The next entity's rank should be adjusted to account for the shared rank among the previous entities. For instance, if two entities share rank 2, the next entity should be assigned rank 4 (since there are two entities with rank 2).

> 3. Set the entities `share_ratio` based on their position in the `rank_table` specified in the recurring transfer.

Reviewer 3: Is there a need for a param for maximum rank table size? If not we should specify the max size that needs to be supported here. I guess it could be pretty big like a few hundred or low thousands entries without issue… can't imagine anyone want tons of entries really.

Author: Good point, I think a hard limit of 500 rows should give more than enough room.

- $s_{i}$ be the share of the rewards for entity $i$

$$d_{i}=r_{i} M_{i}$$

Note if the entity is a team, $M_{i}$ is set to `1` as reward payout multipliers are considered later when distributing rewards [amongst the team members](#distributing-rewards-amongst-team-members).

Calculate each entities share of the rewards, $s_{i}$ pro-rata based on $d_{i}$, i.e.

$$s_{i} = \frac{d_{i}}{\sum_{i=1}^{n}d_{i}}$$

#### Distributing with exponential decay

Rewards funded using the exponential-decay strategy should be distributed as follows.

1. Calculate each entities reward metric
2. Order each entity in a descending list by their reward metric value and determine their "rank" in the list
3. Normalise the rank of each party within a range of 0 to 1, using the highest rank as the upper limit for normalisation.
4. Calculate each entities share of the rewards using the below formula.

Let:

- $d_{i}$ be the payout factor for entity $i$
- $s_{i}$ be the share of the rewards for entity $i$

> - $k$ be the decay factor specified in the recurring transfer funding the reward. This must be a float in the range $0\leq k \leq 2$. 

Reviewer 2: ```suggestion
- $k$ be the decay factor specified in the recurring transfer funding the reward. This must be a decimal in the range $0\leq k \leq 2$. 
```

> - $R_{i}$ be the normalised rank of entity $i$

Reviewer 2: Did we say how it's normalised somehwere?

Author: Step 3 on line 227
> 3. Normalise the rank of each party within a range of 0 to 1, using the highest rank as the upper limit for normalisation.

- $M_{i}$ be the sum of all reward payout multipliers for entity $i$ (reward payout multipliers include the [activity streak multiplier](./0086-ASPR-activity_streak_program.md#applying-the-activity-reward-multiplier) and [bonus rewards multiplier](./0085-RVST-rewards_vesting.md#determining-the-rewards-bonus-multiplier)).

$$d_{i}=M_{i} e^{-k R_{i}}$$

Note if the entity is a team, $M_{i}$ is set to 1 as reward payout multipliers are considered later when distributing rewards [amongst the team members](#distributing-rewards-amongst-team-members).

> To avoid exponential formulas in the core implementation, the above exponential equation can be approximated using the following 7th order Taylor expansion. An odd number of terms has intentionally been chosen so the expansion diverges to $-\infty$ rather than $+\infty$.

Reviewer 2: ```suggestion
To avoid exponential formulas in the core implementation, the above exponential function can be approximated using the following 7th order Taylor expansion. An odd number of terms has intentionally been chosen so the expansion diverges to $-\infty$ rather than $+\infty$.
```

Given:
    rank_table = [
        {"start_rank": 1, "share_ratio": 10},
        {"start_rank": 2, "share_ratio": 5},
        {"start_rank": 4, "share_ratio": 2},
        {"start_rank": 10, "share_ratio": 1},
        {"start_rank": 20, "share_ratio": 0},
    ]
    rank=6

Then:
    share_ratio=2
```

Calculate each entities share of the rewards as follows.

Let:

- $d_{i}$ be the payout factor for entity $i$
- $s_{i}$ be the share of the rewards for entity $i$
- $k$ be the decay factor specified in the recurring transfer funding the reward. This must be a float in the range $0\leq k \leq 2$. 
- $R_{i}$ be the normalised rank of entity $i$
- $M_{i}$ be the sum of all reward payout multipliers for entity $i$ (reward payout multipliers include the [activity streak multiplier](./0086-ASPR-activity_streak_program.md#applying-the-activity-reward-multiplier) and [bonus rewards multiplier](./0085-RVST-rewards_vesting.md#determining-the-rewards-bonus-multiplier)).

$$d_{i}=M_{i} * r_{i}$$

Note if the entity is a team, $M_{i}$ is set to 1 as reward payout multipliers are considered later when distributing rewards [amongst the team members](#distributing-rewards-amongst-team-members).

> To avoid exponential formulas in the core implementation, the above exponential equation can be approximated using the following 7th order Taylor expansion. An odd number of terms has intentionally been chosen so the expansion diverges to $-\infty$ rather than $+\infty$.

Reviewer 3: Should this be gone?

> $$d_{i} = M_{i} e^{-k R_i} \approx M_{i} \min(\sum_{j=0}^{7} \frac{(-k R_i)^j}{j!}, 0)$$

Reviewer 3: Should this be gone too?
